File: dcp-solutions/2024-10-23/solution.go
```go
package main

import (
	"fmt"
)

func countInversions(A []int) (int, int) {
	inversionCount := 0
	var merge func(left, right []int)
	merge = func(left, right []int) {
		merged := make([]int, len(left)+len(right))
		leftIndex := 0
		rightIndex := 0
		mergedIndex := 0

		for leftIndex < len(left) && rightIndex < len(right) {
			if left[leftIndex] <= right[rightIndex] {
				merged[mergedIndex] = left[leftIndex]
				leftIndex++
			} else {
				merged[mergedIndex] = right[rightIndex]
				rightIndex++
				inversionCount += len(left) - leftIndex
			}
			mergedIndex++
		}

		for leftIndex < len(left) {
			merged[mergedIndex] = left[leftIndex]
			leftIndex++
			mergedIndex++
		}

		for rightIndex < len(right) {
			merged[mergedIndex] = right[rightIndex]
			rightIndex++
			mergedIndex++
		}

		return inversionCount, merged
	}

	var mergeSort func(left, right []int)
	mergeSort = func(left, right []int) (inversionCount int, merged []int) {
		if len(left) == 0 {
			return inversionCount, make([]int, 0)
		}
		if len(right) == 0 {
			return inversionCount, left
		}

		if len(left) < len(right) {
			leftInversion, merged = mergeSort(left, right[:len(left)])
			return leftInversion + merged
		} else {
			rightInversion, merged = mergeSort(left[:len(right)], right)
			return merged + rightInversion
		}
	}

	inversionCount, _ = mergeSort(A, make([]int, 0))
	return inversionCount, A
}

func main() {

}
```

File: dcp-solutions/2024-10-23/solution.py
```python
def count_inversions(array):
    def merge_sort(array):
        if len(array) <= 1:
            return array, 0
        mid = len(array) // 2
        left, left_inv = merge_sort(array[:mid])
        right, right_inv = merge_sort(array[mid:])
        merged, inv = merge(left[0], right[0])
        left_inv += right_inv
        return merged, left_inv + right_inv

    def merge(left, right):
        result = []
        i, j = 0, 0
        inv_count = 0
        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                result.append(left[i])
                i += 1
            else:
                result.append(right[j])
                j += 1
                inv_count += len(left) - i
        result.extend(left[i:])
        result.extend(right[j:])
        return result, inv_count

    return merge_sort(array)[1]
```